### PR TITLE
Exceptions

### LINKED ISSUE #182: Request for comments: Unify exception throwing
We're currently throwing `ObjectBoxException` as well as `Exception`. 

* Either get rid of the former or use it in all places? 
* Or actually use more specific classes, depending on what kind of exception it is? Like `ArgumentError`, `StateError` and reserve the generic `Exception` for cases when we can't do any better?
* What's the best practice?

### PR DESCRIPTION
**Pull request by vaind**
closes #182

### CODE REVIEW

File: objectbox/lib/src/modelinfo/modelproperty.dart [hunk 1]
```diff
@@ -23,7 +23,7 @@ class ModelProperty {
 
   set name(String? value) {
     if (value == null || value.isEmpty) {
-      throw Exception('name must not be null or an empty string');
+      throw ArgumentError('name must not be null or an empty string');
```
Comment (greenrobot-team) [inline objectbox/lib/src/modelinfo/modelproperty.dart:26]: Maybe use `ArgumentError.value(...)` in this file, too? But I guess it just affects formatting of the output and does not provide source code linking or similar.

File: objectbox/lib/src/native/sync.dart [hunk 2]
```diff
@@ -554,7 +554,7 @@ class Sync {
   /// Create a Sync annotation, enabling synchronization for an entity.
   const Sync();
 
-  static late final bool _syncAvailable = C.sync_available();
+  static late final bool _syncAvailable = C.has_feature(OBXFeature.Sync);
```
Comment (greenrobot-team) [inline objectbox/lib/src/native/sync.dart:557]: I guess this just snuck in here? Fine to keep, just double checking if not by accident.

File: objectbox/lib/src/relations/to_many.dart [hunk 3]
```diff
@@ -243,7 +244,7 @@ class ToMany<EntityT> extends Object with ListMixin<EntityT> {
 
   void _verifyAttached() {
     if (!_attached) {
-      throw Exception('ToMany relation field not initialized. '
+      throw ObjectBoxException('ToMany relation field not initialized. '
```
Comment (greenrobot-team) [inline objectbox/lib/src/relations/to_many.dart:247]: Why not `StateError`?

File: objectbox/lib/src/relations/to_one.dart [hunk 4]
```diff
@@ -142,7 +143,7 @@ class ToOne<EntityT> {
 
   void _verifyAttached() {
     if (!_attached) {
-      throw Exception('ToOne relation field not initialized. '
+      throw ObjectBoxException('ToOne relation field not initialized. '
```
Comment (greenrobot-team) [inline objectbox/lib/src/relations/to_one.dart:146]: Same as to-many: why not `StateError`?

**Review by greenrobot-team — COMMENTED** [review]
Found some minor things, LGTM otherwise.

File: objectbox/lib/src/native/sync.dart (hunk 2 above)
Comment (vaind) [reply objectbox/lib/src/native/sync.dart:557]: yeah, noticed the use of the deprecated function so just changed it in the meantime

File: objectbox/lib/src/relations/to_many.dart (hunk 3 above)
Comment (vaind) [reply objectbox/lib/src/relations/to_many.dart:247]: done

File: objectbox/lib/src/relations/to_one.dart (hunk 4 above)
Comment (vaind) [reply objectbox/lib/src/relations/to_one.dart:146]: done

File: objectbox/lib/src/modelinfo/modelproperty.dart (hunk 1 above)
Comment (vaind) [reply objectbox/lib/src/modelinfo/modelproperty.dart:26]: Nah, AFAIK it just concatenates the string. This is fairly internal error anyway so as long as the model isn't messed up, should never really happen

**Pull request merged**